File: libs/go/pkg/nodes_qsort_sorting.go
package sums

type nodesComparer struct {
	pairsNotInPathCache []int
	path                *Path
}

func newNodesComparer(path *Path, maxN int) *nodesComparer {
	return &nodesComparer{
		path:                path,
		pairsNotInPathCache: make([]int, maxN+1),
	}
}

func (rcvr *nodesComparer) FlushCache() {
	if rcvr.path == nil {
		return
	}
	for i, _ := range rcvr.pairsNotInPathCache {
		rcvr.pairsNotInPathCache[i] = -1
	}
}

func (rcvr *nodesComparer) getPairsNotInPath(node *Node) int {
	result := rcvr.pairsNotInPathCache[node.value]
	if result == -1 {
		result = node.PairsNotInPathCount(rcvr.path)
		rcvr.pairsNotInPathCache[node.value] = result
	}
	return result
}

func (rcvr *nodesComparer) Compare(nodeA *Node, nodeB *Node) int {
	if rcvr.path != nil {
		a := rcvr.getPairsNotInPath(nodeA)
		b := rcvr.getPairsNotInPath(nodeB)

		if a != b {
			if a < b {
				return -1
			}
			if a > b {
				return 1
			}
		}
	}

	a := nodeA.PairsCount()
	b := nodeB.PairsCount()

	if a != b {
		if a < b {
			return -1
		}
		if a > b {
			return 1
		}
	}

	a = nodeA.Value()
	b = nodeB.Value()

	if a > b {
		return -1
	}
	if a < b {
		return 1
	}
	return 0
}

func partition(comparer *nodesComparer, nodes []*Node, start int, end int) int {
	var temp *Node
	marker := start
	for i := start; i < end; i++ {
		var compareResult = comparer.Compare(nodes[i], nodes[end])
		if compareResult == -1 {
			temp = nodes[marker]
			nodes[marker] = nodes[i]
			nodes[i] = temp
			marker += 1
		}
	}

	temp = nodes[marker]
	nodes[marker] = nodes[end]
	nodes[end] = temp
	return marker
}

func quicksort(comparer *nodesComparer, nodes []*Node, start int, end int) {
	if start >= end {
		return
	}

	pivot := partition(comparer, nodes, start, end)
	quicksort(comparer, nodes, start, pivot-1)
	quicksort(comparer, nodes, pivot+1, end)
}

type NodesSortingQSort struct {
	comparer *nodesComparer
}

func NewNodesSortingQSort(path *Path, maxN int) *NodesSortingQSort {
	return &NodesSortingQSort{
		comparer: newNodesComparer(path, maxN),
	}
}

func (rcvr *NodesSortingQSort) SortNodes(nodes []*Node) {
	rcvr.comparer.FlushCache()
	quicksort(rcvr.comparer, nodes, 0, len(nodes)-1)
}

File: libs/rust/square-sums/src/nodes_qsort_sorting.rs
use std::{
    cmp::Ordering,
    rc::{Rc, Weak},
};

use crate::{
    node::Node,
    nodes_compare::{nodes_comparer, nodes_comparer_with_path},
    pairs_not_in_path_cache::PairsNotInPathCache,
    path::Path,
};

#[cfg(feature = "qsort")]
fn partition<N, C: FnMut(&N, &N) -> Ordering>(
    comparer: &mut C,
    nodes: &mut [N],
    start: usize,
    end: usize,
) -> usize {
    let mut marker = start;
    for i in start..end {
        let compare_result = comparer(&nodes[i], &nodes[end]);
        if compare_result == Ordering::Less {
            nodes.swap(i, marker);
            marker += 1;
        }
    }

    nodes.swap(marker, end);
    marker
}

#[cfg(feature = "qsort")]
fn quicksort<N, C: FnMut(&N, &N) -> Ordering>(mut comparer: C, nodes: &mut [N]) {
    quicksort_recursive(&mut comparer, nodes, 0, nodes.len() - 1)
}

#[cfg(feature = "qsort")]
fn quicksort_recursive<N, C: FnMut(&N, &N) -> Ordering>(
    comparer: &mut C,
    nodes: &mut [N],
    start: usize,
    end: usize,
) {
    if start >= end {
        return;
    }

    let pivot = partition(comparer, nodes, start, end);
    if pivot > 0 {
        quicksort_recursive(comparer, nodes, start, pivot - 1);
    }
    quicksort_recursive(comparer, nodes, pivot + 1, end);
}

#[cfg(feature = "qsort")]
pub struct QSortNodesSorting {}

#[cfg(feature = "qsort")]
impl QSortNodesSorting {
    pub fn new() -> QSortNodesSorting {
        QSortNodesSorting {}
    }

    pub fn sort_nodes(&self, nodes: &mut [Rc<Node>]) {
        quicksort(|i, j| nodes_comparer(i, j).expect(""), nodes);
    }
}

#[cfg(feature = "qsort")]
pub struct QSortNodesSortingWithCache {
    cache: PairsNotInPathCache,
}

#[cfg(feature = "qsort")]
impl QSortNodesSortingWithCache {
    pub fn new(n: u32) -> QSortNodesSortingWithCache {
        QSortNodesSortingWithCache {
            cache: PairsNotInPathCache::new(n),
        }
    }

    pub fn sort_nodes(&mut self, path: &Path, nodes: &mut [Weak<Node>]) {
        self.cache.flush();
        quicksort(
            |i, j| nodes_comparer_with_path(&mut self.cache, path, i, j).expect(""),
            nodes,
        );
    }
}
